plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { KubeConfig, bufferFromFileOrString } from '@kubernetes/client-node';
+import { Logger } from 'winston';
+import fetch from 'node-fetch';
+import * as https from 'https';
+
+import type { Request } from 'express';
+
+import { ClusterDetails, KubernetesClustersSupplier } from '../types/types';
+
+const HEADER_CONTENT_TYPE: string = 'Content-Type';
+const APPLICATION_JSON: string = 'application/json';
+
+const HEADER_KUBERNETES_CLUSTERS: string = 'X-Kubernetes-Clusters';
+
+const ERROR_BAD_REQUEST: number = 400;
+const ERROR_NOT_FOUND: number = 404;
+const ERROR_INTERNAL_SERVER: number = 500;
+
+export interface KubernetesProxyServices {
+  kcs: KubernetesClustersSupplier;
+}
+
+export interface KubernetesProxyResponse {
+  code: number;
+  data: any;
+  cluster?: string;
+}
+
+interface KubernetesProxyClusters {
+  [key: string]: string;
+}
+
+export class KubernetesProxy {
+  constructor(protected readonly logger: Logger) {}
+
+  public async handleProxyRequest(
+    services: KubernetesProxyServices,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const krc = this.getKubernetesRequestedClusters(req);
+    const details = await this.getClusterDetails(services.kcs, krc);
+
+    if (details.length < 1) {
+      this.logger.error(
+        'no clusters found for request - check the app config file to make sure all clusters are registered in Backstage',
+      );
+      return {
+        code: ERROR_NOT_FOUND,
+        data: 'No clusters found!',
+      };
+    }
+
+    const responses = await Promise.all(
+      details.map(async d => {
+        const response = await this.makeRequestToCluster(d, req);
+        return response;
+      }),
+    );
+
+    const data: { [key: string]: any } = {};
+    const codes: number[] = [];
+
+    responses.forEach(kpr => {
+      if (kpr.cluster) {
+        data[kpr.cluster] = kpr.data;
+        codes.push(kpr.code);
+      }
+    });
+
+    const code = this.getBestResponseCode(codes);
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  protected getKubernetesRequestedClusters(
```

Comment:
i think same thing here, all of this should just be private instead of protected most likely